lgc/elfLinker/FetchShader.cpp

```diff
@@ -114,42 +136,104 @@ Module *FetchShader::generate() {
-          break;
-        }
-        case static_cast<unsigned>(UserDataMapping::BaseVertex):
-          replacement = fetchFunc->getArg(m_vsEntryRegInfo.baseVertex);
-          break;
-        case static_cast<unsigned>(UserDataMapping::BaseInstance):
-          replacement = fetchFunc->getArg(m_vsEntryRegInfo.baseInstance);
-          break;
-        case static_cast<unsigned>(ShaderInput::VertexId):
-          builder.SetInsertPoint(&*fetchFunc->front().getFirstInsertionPt());
-          replacement = builder.CreateBitCast(fetchFunc->getArg(m_vsEntryRegInfo.sgprCount + m_vsEntryRegInfo.vertexId),
-                                              builder.getInt32Ty());
-          break;
-        case static_cast<unsigned>(ShaderInput::InstanceId):
-          builder.SetInsertPoint(&*fetchFunc->front().getFirstInsertionPt());
-          replacement = builder.CreateBitCast(
-              fetchFunc->getArg(m_vsEntryRegInfo.sgprCount + m_vsEntryRegInfo.instanceId), builder.getInt32Ty());
-          break;
-        default:
-          llvm_unreachable("Unexpected special user data or shader input");
-        }
+        replacement = getReplacementForInputBuiltIn(call);
         call->replaceAllUsesWith(replacement);
         call->eraseFromParent();
       }
     }
   }
+}
 
-  return fetchFunc->getParent();
+// =====================================================================================================================
+// @param call : A call to a shader input builtin that needs to be replaced.
+// @returns : The value that is represented by |call|.  It will be in a position where in can be used in place of all
+// uses of |call|.
+Value *FetchShader::getReplacementForInputBuiltIn(CallInst *call) const {
+  switch (cast<ConstantInt>(call->getArgOperand(0))->getZExtValue()) {
+  case static_cast<unsigned>(UserDataMapping::VertexBufferTable):
+    return getReplacementForVertexBufferTableBuiltIn(call);
+  case static_cast<unsigned>(UserDataMapping::BaseVertex):
+    return call->getParent()->getParent()->getArg(m_vsEntryRegInfo.baseVertex);
+  case static_cast<unsigned>(UserDataMapping::BaseInstance):
+    return call->getParent()->getParent()->getArg(m_vsEntryRegInfo.baseInstance);
+  case static_cast<unsigned>(ShaderInput::VertexId):
+    return getReplacementForVertexIdBuiltIn(call);
+  case static_cast<unsigned>(ShaderInput::InstanceId):
+    return getReplacementForInstanceIdBuildIn(call);
+  default:
+    llvm_unreachable("Unexpected special user data or shader input");
+  }
+  return nullptr;
+}
+
+// =====================================================================================================================
+// @param call : A call to the InstanceId shader input builtin.
+// @returns : The value of the InstanceId.  All new code will be place at the start of the function containing call.
+Value *FetchShader::getReplacementForInstanceIdBuildIn(CallInst *call) const {
```

Comment:
Nit BuildIn vs BuiltIn (maybe this is deliberate though?)

Reply:
Fixed.